File: app/src/main/java/com/huanchengfly/tieba/post/fragments/PhotoViewFragment.java
package com.huanchengfly.tieba.post.fragments;

import android.annotation.SuppressLint;
import android.content.DialogInterface;
import android.net.Uri;
import android.os.Bundle;
import android.view.View;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.davemorrissey.labs.subscaleview.SubsamplingScaleImageView;
import com.github.piasy.biv.loader.ImageLoader;
import com.github.piasy.biv.view.BigImageView;
import com.github.piasy.biv.view.GlideImageViewFactory;
import com.huanchengfly.tieba.post.R;
import com.huanchengfly.tieba.post.components.CircleProgressIndicator;
import com.huanchengfly.tieba.post.models.PhotoViewBean;
import com.huanchengfly.tieba.post.utils.DialogUtil;
import com.huanchengfly.tieba.post.utils.ImageUtil;

import java.io.File;

import butterknife.BindView;

public class PhotoViewFragment extends BaseFragment {
    public static final String TAG = PhotoViewFragment.class.getSimpleName();
    private static final String ARG_INFO = "info";
    @BindView(R.id.big_image_view)
    public BigImageView bigImageView;
    private PhotoViewBean photoViewBean;

    public PhotoViewFragment() {
    }

    public static PhotoViewFragment newInstance(PhotoViewBean photoViewBean) {
        PhotoViewFragment fragment = new PhotoViewFragment();
        Bundle args = new Bundle();
        args.putParcelable(ARG_INFO, photoViewBean);
        fragment.setArguments(args);
        return fragment;
    }

    public PhotoViewBean getPhotoViewBean() {
        return photoViewBean;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            photoViewBean = getArguments().getParcelable(ARG_INFO);
        }
    }

    @Override
    public int getLayoutId() {
        return R.layout.fragment_photo_view;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        if (!canLoad()) return;
        loadByBigImageView();
    }

    private void showBottomBar(boolean autoHide) {
        if (getAttachContext() instanceof OnChangeBottomBarVisibilityListener) {
            ((OnChangeBottomBarVisibilityListener) getAttachContext()).onShow(autoHide);
        }
    }

    @SuppressLint("ClickableViewAccessibility")
    private void loadByBigImageView() {
        if (!canLoad()) return;
        bigImageView.setVisibility(View.VISIBLE);
        bigImageView.setImageViewFactory(new GlideImageViewFactory());
        bigImageView.setProgressIndicator(new CircleProgressIndicator());
        bigImageView.showImage(Uri.parse(photoViewBean.getOriginUrl()));
        bigImageView.setOnTouchListener((view, event) -> {
            showBottomBar(true);
            return false;
        });
        bigImageView.setOnLongClickListener(v -> {
            showBottomBar(false);
            openDialog(photoViewBean);
            return true;
        });
        bigImageView.setOnClickListener(view -> {
            if (getActivity() != null) {
                getActivity().finish();
            }
        });
        bigImageView.setImageLoaderCallback(new ImageLoader.Callback() {
            @Override
            public void onCacheHit(int imageType, File image) {
            }

            @Override
            public void onCacheMiss(int imageType, File image) {
            }

            @Override
            public void onStart() {
            }

            @Override
            public void onProgress(int progress) {
            }

            @Override
            public void onFinish() {
            }

            @Override
            public void onSuccess(File image) {
                final SubsamplingScaleImageView view = bigImageView.getSSIV();

                if (view != null) {
                    view.setMinimumDpi(80);

                    view.setOnImageEventListener(new SubsamplingScaleImageView.OnImageEventListener() {
                        @Override
                        public void onReady() {
                        }

                        @Override
                        public void onImageLoaded() {
                            view.setDoubleTapZoomDpi(80);
                            view.setDoubleTapZoomDuration(200);
                            view.setDoubleTapZoomStyle(SubsamplingScaleImageView.ZOOM_FOCUS_FIXED);
                            view.setQuickScaleEnabled(false);
                        }

                        @Override
                        public void onPreviewLoadError(Exception e) {
                        }

                        @Override
                        public void onImageLoadError(Exception e) {
                        }

                        @Override
                        public void onTileLoadError(Exception e) {
                        }

                        @Override
                        public void onPreviewReleased() {
                        }
                    });
                }
            }

            @Override
            public void onFail(Exception error) {
                Toast.makeText(getAttachContext(), R.string.toast_load_failed, Toast.LENGTH_SHORT).show();
            }
        });
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
    }

    private boolean canLoad() {
        return getActivity() != null && !getActivity().isDestroyed() && !getActivity().isFinishing();
    }

    private void openDialog(PhotoViewBean photoViewBean) {
        if (!canLoad()) {
            return;
        }
        final String[] strArray = new String[]{getAttachContext().getString(R.string.menu_save_photo)};
        DialogUtil.build(getAttachContext())
                .setItems(strArray, (DialogInterface dialog, int which) -> {
                    switch (which) {
                        case 0:
                            ImageUtil.download(getAttachContext(), photoViewBean.getOriginUrl(), photoViewBean.isGif());
                            break;
                    }
                })
                .create()
                .show();
    }

    public interface OnChangeBottomBarVisibilityListener {
        void onShow(boolean autoHide);

        void onHide();
    }
}

File: app/src/main/java/com/huanchengfly/tieba/post/fragments/PhotoViewFragment.kt
package com.huanchengfly.tieba.post.fragments

import android.annotation.SuppressLint
import android.content.DialogInterface
import android.os.Bundle
import android.view.View
import butterknife.BindView
import com.github.panpf.sketch.displayImage
import com.github.panpf.sketch.viewability.showRingProgressIndicator
import com.github.panpf.sketch.zoom.SketchZoomImageView
import com.huanchengfly.tieba.post.R
import com.huanchengfly.tieba.post.dpToPx
import com.huanchengfly.tieba.post.dpToPxFloat
import com.huanchengfly.tieba.post.models.PhotoViewBean
import com.huanchengfly.tieba.post.toastShort
import com.huanchengfly.tieba.post.ui.common.theme.utils.ThemeUtils
import com.huanchengfly.tieba.post.utils.DialogUtil
import com.huanchengfly.tieba.post.utils.ImageUtil

class PhotoViewFragment : BaseFragment() {
    @JvmField
    @BindView(R.id.big_image_view)
    var bigImageView: SketchZoomImageView? = null
    var photoViewBean: PhotoViewBean? = null
        private set

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (arguments != null) {
            photoViewBean = arguments!!.getParcelable(ARG_INFO)!!
        }
    }

    override fun getLayoutId(): Int {
        return R.layout.fragment_photo_view
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        if (!canLoad()) return
        loadByBigImageView()
    }

    private fun showBottomBar(autoHide: Boolean) {
        if (attachContext is OnChangeBottomBarVisibilityListener) {
            (attachContext as OnChangeBottomBarVisibilityListener).onShow(autoHide)
        }
    }

    @SuppressLint("ClickableViewAccessibility")
    private fun loadByBigImageView() {
        if (!canLoad()) return
        var url = photoViewBean!!.originUrl
        if (url == null) {
            url = photoViewBean!!.url
        }
        if (url == null) {
            return
        }
        bigImageView!!.visibility = View.VISIBLE
        bigImageView!!.showRingProgressIndicator(48.dpToPx(), 2f.dpToPxFloat(), ThemeUtils.getColorByAttr(attachContext, R.attr.colorPrimary))
        bigImageView!!.displayImage(url) {
            lifecycle(viewLifecycleOwner.lifecycle)
            listener(
                onError = { _, _ ->
                    attachContext.toastShort(R.string.toast_load_failed)
                }
            )
        }
        bigImageView!!.setOnTouchListener { _, _ ->
            showBottomBar(true)
            false
        }
        bigImageView!!.setOnLongClickListener {
            showBottomBar(false)
            openDialog(photoViewBean)
            true
        }
        bigImageView!!.setOnClickListener {
            if (activity != null) {
                activity!!.finish()
            }
        }
    }

    private fun canLoad(): Boolean {
        return activity != null && !activity!!.isDestroyed && !activity!!.isFinishing
    }

    private fun openDialog(photoViewBean: PhotoViewBean?) {
        if (!canLoad()) {
            return
        }
        val strArray = arrayOf(attachContext.getString(R.string.menu_save_photo))
        DialogUtil.build(attachContext)
            .setItems(strArray) { _: DialogInterface?, which: Int ->
                when (which) {
                    0 -> ImageUtil.download(attachContext, photoViewBean!!.originUrl)
                }
            }
            .create()
            .show()
    }

    interface OnChangeBottomBarVisibilityListener {
        fun onShow(autoHide: Boolean)
        fun onHide()
    }

    companion object {
        val TAG: String = PhotoViewFragment::class.java.simpleName
        private const val ARG_INFO = "info"

        @JvmStatic
        fun newInstance(photoViewBean: PhotoViewBean?): PhotoViewFragment {
            val fragment = PhotoViewFragment()
            val args = Bundle()
            args.putParcelable(ARG_INFO, photoViewBean)
            fragment.arguments = args
            return fragment
        }
    }
}